src/shared/sam/activation.ts

```diff
@@ -27,13 +30,55 @@ import { SamDebugConfigProvider } from './debugger/awsSamDebugger'
 import { addSamDebugConfiguration } from './debugger/commands/addSamDebugConfiguration'
 import { SamDebugSession } from './debugger/samDebugSession'
 import { AWS_SAM_DEBUG_TYPE } from './debugger/awsSamDebugConfiguration'
+import { ext } from '../extensionGlobals'
+
+const YAML_PLUGIN = 'redhat.vscode-yaml'
+const YAML_PROMPT = 'dontShowYamlPromptEver'
 
 /**
  * Activate SAM-related functionality.
  */
 export async function activate(ctx: ExtContext): Promise<void> {
     initializeSamCliContext({ settingsConfiguration: ctx.settings })
 
+    const dontPromptEverAgain = ext.context.globalState.get<boolean>(YAML_PROMPT)
+    // only pop this up in VS Code and Insiders since other VS Code-like IDEs (e.g. Theia) may not have a marketplace or contain the YAML plugin
+    const isVsCode = vscode.env.appName && vscode.env.appName.startsWith('Visual Studio Code') ? true : false
+
+    if (!dontPromptEverAgain && !vscode.extensions.getExtension(YAML_PLUGIN) && isVsCode) {
+        // these will all be disposed immediately after showing one so the user isn't prompted more than once per session
+        const yamlPromptDisposables: vscode.Disposable[] = []
+
+        vscode.workspace.onDidOpenTextDocument(
+            async (doc: vscode.TextDocument) => {
+                if (doc.fileName.endsWith('template.yaml') || doc.fileName.endsWith('template.yml')) {
+                    promptInstallYamlPlugin()
+                    for (const prompt of yamlPromptDisposables) {
+                        prompt.dispose()
+                    }
+                }
+            },
+            undefined,
+            yamlPromptDisposables
+        )
+
+        vscode.window.onDidChangeActiveTextEditor(
```

Comment:
Seems reasonable to do the check when opening a doc. No point in re-checking after a doc is open and the user switches active documents.

Reply:
It's for if a user already has multiple docs open, activates the extension, and navigates to their template file. It still only triggers once between the two events.

The only time it won't show up is if you have focus on a template.yaml file and activate the extension. At that point, it'll only show on close and reopen or refocus.

Reply:
With the yaml files auto-activating the extension, is that still necessary?
Its a great scenario. If it is required, could the code be de-duplicated (with both onXxx events) somehow?

Reply:
Yes because by the time the extension has already loaded, VS Code has already sent the event, unless you want me to do a one-time "is the focused doc a template" on extension load.

re: deduping logic (also answering @justinmk3 's question below), I'm just afraid of the scoping for `yamlPromptDisposables` if this goes to a separate function. I can give it another look though, I guess the array won't be mutated and the disposables contained within should still be disposed of correctly.

Reply:
> I'm just afraid of the scoping for `yamlPromptDisposables` if this goes to a separate function.

the function can be local to another function i.e. the same scope as `yamlPromptDisposables`

Reply:
Consider adding a reusable `once()` function for these. Seems like a common concern and avoid the array mutation oddness. It immediately disposes its subscription once fired and ignores all future fires.

https://github.com/microsoft/vscode/blame/fbb4aaa499fef9fbabb6f43bbf2db460ef63b2f3/src/vs/base/common/event.ts#L26